_posts/2020-01-09-ios-architectures.md

```diff
@@ -0,0 +1,1029 @@
+
+```
+protocol DetailViewInterface: class {
+    func setMessage(message: String)
+}
+```
+
+And likewise the view itself:
+
+```
+import UIKit
+
+final class DetailViewController: UIViewController {
+
+    let eventHandler: DetailPresenter
+
+    weak var label: UILabel?
+    weak var button: UIButton?
+
+    init(eventHandler: DetailPresenter) {
+       self.eventHandler = eventHandler
+       super.init(nibName: nil, bundle: nil)
+   }
+
+   required init?(coder: NSCoder) {
+       fatalError("coder: init not supported")
+   }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        title = "Detail"
+
+        view.backgroundColor = UIColor.blue
+        setupLabel()
+        setupCancelButton()
+
+        eventHandler.loadData()
+    }
+
+
+    @objc func dismissAction() {
+        eventHandler.dismiss()
+    }
+
+    private func setupLabel() {
+        let label = UILabel(frame: CGRect(x: 0, y: 250, width: 350, height: 55))
+        label.textAlignment = .center
+        label.textColor = UIColor.white
+        view.addSubview(label)
+        self.label = label
+    }
+
+    private func setupCancelButton() {
+        let button = UIButton(frame: CGRect(x: 0, y: 350, width: 350, height: 55))
+        button.setTitle("Cancel", for: .normal)
+        button.setTitleColor(.white, for: .normal)
+        button.addTarget(self, action: #selector(dismissAction), for: .touchUpInside)
+        view.addSubview(button)
+        self.button = button
+    }
+}
+```
+
+Finally we implement the protocol:
+
+```
+extension DetailViewController: DetailViewInterface {
+    func setMessage(message: String) {
+        label?.text = message
+    }
+}
+```
+
+Note how most of the view's code in both cases is set up the various subviews and link the actions that the user can perform. That's the whole range of responsibilities of the view. The presenter will convert that user intent to an action to dispatch to either the interactor or the wireframe and those two will act.
+
+You can find the full project on [Github][1].
+
+## MVP-R [Coming Soon]
+
+<!--
+
+MVP is the simplest one of the alternatives (exclusing MVC), its biggest issue is not providing routing capabilities so the Presenter on top of handling User Interaction and communicating with the netork service/persistance layer has to handle the navigation flow as well. With the addition of a Router we solve this problem.
+
+The ViewController acts as a view that gets esposed to the presenter as a protocol (as it happens in VIPER). The Presenter handles the business logic and dispatches the changes back to the view (controller) for display. The router handles the navigation.
+
+ TODO 
+ 
+ -->
+
+## MVC [Coming Soon]
+
+<!-- TODO -->
+
+## What should I use?
+
+You should not just use of architecture and stick with it no matter what. What architecture to use depends heavily on what kind of application you're building and what it is trying to make.
```

Comment:
use one - - //of//--

Reply:
to make or to do?  If think apps make much

Reply:
do =)